### PR TITLE
Make the functions async

### PR DESCRIPTION
**Pull request by elpiel**

### CODE REVIEW

**Comment by elpiel** [comment]
Ok so CircleCi should be improved as now it will install `nightly` every day. We should probably also freeze the nightly version, so that we don't have problems with it.

**Comment by DavidBM** [comment]
Great! 
Async/Await will come out of beta very very soon. Once out we should have the CI using the stable

**Comment by elpiel** [comment]
It depends. The crates are a different story than Rust. It depends when will `tokio` and other crates like that go out to stable.

**Comment by elpiel** [comment]
@DavidBM you can check the latest changes. I did move the test of `unzip_archive` to `tests` folder as it looks like an integration test to me as it uses a real file and we do have it public.
We do need to think of a way to make sure the `temp` (or wherever we decide to store the temporary files) is cleaned.

Keep in mind - everything can be change if you don't like naming and etc.

File: Cargo.toml [hunk 1]
```diff
@@ -8,8 +8,9 @@ edition = "2018"
 clap = { version = "2.33.0", features = ["yaml"] }
 zip = "0.5.3"
 ignore = "0.4"
-globset = "0.3"
-
+# globset = "0.3"
```
Comment (elpiel) [inline Cargo.toml]: commented it out until we enable this and uncomment the code in `options` module

File: src/lib.rs [hunk 2]
```diff
@@ -1,121 +1,76 @@
...
-    let mut archive = zip::ZipArchive::new(file).unwrap();
+pub async fn unzip_archive(file: File, directory: &Path) -> Result<(), Box<dyn std::error::Error>> {
+    let mut archive = zip::ZipArchive::new(file)?;
+    let mut files = vec![];
 
     for file_index in 0..archive.len() {
-        let file = archive.by_index(file_index).unwrap();
-        unzip_archive(file, directory)
+        match archive.by_index(file_index) {
+            Ok(file) => files.push(unzip_file(file, directory).await),
```
Comment (elpiel) [inline src/lib.rs]: it does partially solves the problem with mutability and the `by_index`, we can always look into how to improve it later.

File: src/lib.rs [hunk 3]
```diff
@@ -1,121 +1,76 @@
...
-};
+use std::{fs::{File}, io, path::{Path, PathBuf}};
+use async_std::{fs::{create_dir_all, set_permissions, Permissions}};
+use zip::read::ZipFile;
 
-pub fn unzip_file(file: File, directory: &Path) {
-    let mut archive = zip::ZipArchive::new(file).unwrap();
+pub async fn unzip_archive(file: File, directory: &Path) -> Result<(), Box<dyn std::error::Error>> {
+    let mut archive = zip::ZipArchive::new(file)?;
+    let mut files = vec![];
```
Comment (elpiel) [inline src/lib.rs]: this `Vec<Result<...>>` should still be used and dealt with.

File: src/lib.rs [hunk 4]
```diff
@@ -1,121 +1,85 @@
...
+        create_folder(&name, &outpath, directory).await;
     } else {
-        create_file(&name, &outpath, directory, &mut file)
+        create_file(&name, &outpath, directory, &mut file).await;
     }
 
     // Get and Set permissions
-    #[cfg(unix)]
-    set_unix_permissions(&outpath, directory, &file);
+    if cfg!(unix) {
```
Comment (elpiel) [inline src/lib.rs]: I am actually not sure if this or your variant is better.

File: src/lib.rs (hunk 3 above)
Comment (DavidBM) [reply src/lib.rs]: I would call this variable "results" or something like that as it contains the results and not the file itself.

**Comment by DavidBM** [comment]
I think we can merge it. I like the changes.

File: src/lib.rs (hunk 3 above)
Comment (elpiel) [reply src/lib.rs]: Renamed.

**Pull request merged**